perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py

```diff
@@ -121,74 +266,85 @@ def RunNetperf(vm, benchmark_name, server_ip):
-  if unit == MBPS:
-    metric = '%s_Throughput' % benchmark_name
-  else:
-    metric = '%s_Transaction_Rate' % benchmark_name
+  remote_script_path = '%s/run/%s' % (vm.GetScratchDir(), REMOTE_SCRIPT)
+  remote_cmd = '%s --netperf_cmd="%s" --num_streams=%s --port_start=%s' % \
+               (remote_script_path, netperf_cmd, num_streams, PORT_START)
+  remote_stdout, _ = vm.RemoteCommand(remote_cmd)
 
-  meta_keys = [('Confidence Iterations Run', 'confidence_iter'),
-               ('Throughput Confidence Width (%)', 'confidence_width_percent')]
-  metadata = {meta_key: row[np_key] for np_key, meta_key in meta_keys}
-  metadata.update(netperf_test_length=FLAGS.netperf_test_length,
-                  max_iter=FLAGS.netperf_max_iter or 1)
+  # Decode stdouts, stderrs, and return codes from remote command's stdout
+  stdouts, stderrs, return_codes = json.loads(remote_stdout)
 
-  samples = [sample.Sample(metric, value, unit, metadata)]
+  # Metadata to attach to samples
+  metadata = {'netperf_test_length': FLAGS.netperf_test_length,
+              'max_iter': FLAGS.netperf_max_iter or 1}
 
-  # No tail latency for throughput.
-  if unit == MBPS:
-    return samples
+  parsed_output = [_ParseNetperfOutput(stdout, metadata, benchmark_name,
+                                       enable_latency_histograms)
+                   for stdout in stdouts]
 
-  if FLAGS.netperf_enable_histograms:
-    # Generate a sample containing the entire histogram of
-    # latencies.
-    hist = netperf.ParseHistogram(stdout)
-    hist_metadata = {'histogram': json.dumps(hist)}
-    hist_metadata.update(metadata)
-    samples.append(sample.Sample(
-        '%s_Latency_Histogram' % benchmark_name, 0, 'us', hist_metadata))
-
-  for metric_key, metric_name in [
-      ('50th Percentile Latency Microseconds', 'p50'),
-      ('90th Percentile Latency Microseconds', 'p90'),
-      ('99th Percentile Latency Microseconds', 'p99'),
-      ('Minimum Latency Microseconds', 'min'),
-      ('Maximum Latency Microseconds', 'max'),
-      ('Stddev Latency Microseconds', 'stddev')]:
-    samples.append(
-        sample.Sample('%s_Latency_%s' % (benchmark_name, metric_name),
-                      float(row[metric_key]), 'us', metadata))
-  return samples
+  if len(parsed_output) == 1:
+    # Only 1 netperf thread
+    throughput_sample, latency_samples, histogram = parsed_output[0]
+    return [throughput_sample] + latency_samples
+  else:
+    # Multiple netperf threads
+
+    samples = []
+
+    # Unzip parsed output
+    throughput_samples, _, latency_histograms = [list(t)
+                                                 for t in zip(*parsed_output)]
+    # They should all have the same units
+    throughput_unit = throughput_samples[0].unit
+    # Extract the throughput values from the samples
+    throughputs = [s.value for s in throughput_samples]
+    # Compute some stats on the throughput values
+    throughput_stats = sample.PercentileCalculator(throughputs, [50, 90, 99])
```

Comment:
Since we import numpy we might want to look at numpy.percentile http://docs.scipy.org/doc/numpy-1.8.1/reference/generated/numpy.percentile.html to do this calculation .  It doesn't take a histogram but rather an array of the values.